### ISSUE TITLE
UART communication between ev3 (input port1) and arduino is not working

### ISSUE THREAD

**Issue by amreo** [ev3dev/ev3dev#1482]
**What are you trying to do?**
I'm trying to use UART for communicating between arduino and ev3, and print some receiving bytes from the arduino
```
#!/usr/bin/env pybricks-micropython
from pybricks.parameters import Port, Color, Button, Direction, Stop
from pybricks.iodevices import DCMotor, UARTDevice, LUMPDevice, I2CDevice
x = UARTDevice(Port.S1, 9600)
print(x.read_all())
```

**What did you expect to happen**
This simple micropython script prints something.

**What actually happened**
The script prints `b''`

**Source Code**
Here is the program I am trying to run.

(ev3)
```
from time import sleep
brickrun -r -- pybricks-micropython
from pybricks.parameters import Port, Color, Button, Direction, Stop
from pybricks.iodevices import DCMotor, UARTDevice, LUMPDevice, I2CDevice
x = UARTDevice(Port.S1, 9600)
print(x.read_all())
```
(arduino uno)
```
#include <SoftwareSerial.h>

SoftwareSerial Serial1(12, 11); // RX, TX

void setup() {
  //Initialize serial and wait for port to open:
  Serial.begin(115200);
  Serial1.begin(9600);

  while (!Serial) {
    ; // wait for serial port to connect. Needed for native USB port only
  }
}

void loop() {
  Serial.print("A");
  Serial1.print("A");
}
```

**Arduino UNO (please complete the following information):**
 - OS and version: none
 - Connection Type: UART (other-uart)
 - Software: arduino example code

**Robot (please complete the following information):**
 - Device: EV3
 - Versions: ev3dev-stretch-ev3-generic-2020-04-10, 4.14.117-ev3dev-2.3.5-ev3

**System info (from `ev3dev-sysinfo`)**
```
Image file:         ev3dev-stretch-ev3-generic-2020-04-10
Kernel version:     4.14.117-ev3dev-2.3.5-ev3
Brickman:           0.10.3
BogoMIPS:           148.88
Bluetooth:          
Board:              board0
BOARD_INFO_HW_REV=7
BOARD_INFO_MODEL=LEGO MINDSTORMS EV3
BOARD_INFO_ROM_REV=6
BOARD_INFO_SERIAL_NUM=00165352B390
BOARD_INFO_TYPE=main
```

**Background Information**
I am trying to create a UART(and not i2c that works) communication between ev3 and arduino for experimental purpose...
I have also tried to use the LUMP protocol (https://lejosnews.wordpress.com/2014/06/03/arduino-uart-sensors/) with a little more success(the mode and sensor info are detected, reading don't work)

**wire links**
Pin#1(ev3) (white) --- GND(arduino)
Pin#2 (ev3) (black) NC
Pin#3 (ev3) (red) --- GND(arduino) --- Pin#1(ev3) (white)
Pin#4 (ev3) (green) --- 5V (arduino)
Pin#5 (ev3) (yellow) --- D12 (arduino)
Pin#6 (ev3) (blue) --- D11 (arduino)

**Comment by dlech** [comment]
`read_all()` actually means "read what is currently in the buffer, if anything".  Creating the `UARTDevice` probably clears the buffer. So there probably isn't enough time to receive anything new before printing.

Try `x.read(1)` instead to block until it actually receives one byte.

> Pin#4 (ev3) (green) --- 5V (arduino)

Is this actually powering the Arduino? If not, don't connect it.

**Comment by amreo** [comment]
> > Pin#4 (ev3) (green) --- 5V (arduino)
> 
> Is this actually powering the Arduino? If not, don't connect it.

No, Arduino is powered through USB, so I have removed the wire.



> Try `x.read(1)` instead to block until it actually receives one byte.

Done, and return `b'\x05'` insted of `b'A'` or `b'\x41`

**Comment by dlech** [comment]
What does it do if you change it to this?

```python
while True:
    print(x.read(1))
```

**Comment by amreo** [comment]
Print many b'A'... What?

**Comment by amreo** [comment]
@dlech after I have rebooted EV3 (or better I have wrongly shortcuted vcc and gnd) I continue to get b'\x05' instead of b'A'...

**Comment by dlech** [comment]
Maybe a timing issue with the SoftwareSerial on the Arduino? This is the point where I would get out the oscilloscope or logic analyzer to see what is going on.

**Comment by amreo** [comment]
I don't have a oscilloscope or logic analyzer. The output isn't always determistic. I have tried to use the hardware serial but the output isn't the right output and determistic. I suspect is a hardware/connection problem because moving the wire change the output...

**Comment by Oxymore-coder** [comment]
A little late, but, try unplugging the ground wires. They are not required for UART, and having them messes things up on my part.

**Comment by amreo** [comment]
I have tried to unplug the ground wires, but the issue is the same

I have tried to analyze the signal with my (recent) oscilloscope. This is the waveform of D11 pin (with the arduino GND) and with white/red pin attached to ground.

![IMG_008](https://user-images.githubusercontent.com/12671924/208513061-ef9c9775-cd5e-40d9-a377-9f9d272b65a3.png)
![DATA000](https://user-images.githubusercontent.com/12671924/208513068-87339ea9-3317-4964-a31a-b2ccacddcadf.SVG)

Is what you would expect @dlech?

(I don't have updated ev3dev)

**Comment by Oxymore-coder** [comment]
I don't know how to read oscilloscope output, but I can give you the exact code I made it work with:

       EV3:
```
		from pybricks.iodevices import UARTDevice

		arduino = UARTDevice( Port.S1, 9600 )
		while True:
			print(arduino.read(1))
```

        arduino:
```
                void setup() {
		    Serial1.begin(9600);
		}

		void loop() {
  			Serial1.write(0x42);
    			delay(500);
		}
```

Here, the EV3 is plugged in the computer, writing `B` to the VSCode output at a half second interval
The connection can lag, it needs a bit of time to settle

Maybe using SoftwareSerial instead of builtin Serials is somehow messing it up, maybe with different UART configurations, so try `Serial`, `Serial1` or any other like in my code

**Comment by amreo** [comment]
> Maybe using SoftwareSerial instead of builtin Serials is somehow messing it up, maybe with different UART configurations, so try `Serial`, `Serial1` or any other like in my code

Apparently it seems to work (but not using Serial1 because Arduino UNO doesnt't have it and with import Port)

**Comment by amreo** [comment]
But it seems unstable: with 57600 as baudrate and without the delay, it reads '\t' (and serial hw), even if in the doc there is written

> The UART communication functions as asynchronous communication,
supporting from 2400 bit/s to 460 Kbit/s on port 1 & 2, while ports 3 & 4 supports up to 230 Kbit/s.

(Hardware developer kit pdf)

**Comment by Oxymore-coder** [comment]
That could be a limitation on the arduino's side. I rarely go above 9600 with devices that aren't my laptop, that makes it safe, and with short messages, it's not a problem. Don't go too high up and you should be fine
